sage_code/strong_uniform_bounds.py

```diff
@@ -57,19 +56,27 @@ def is_non_increasing(t, part):
     return True
 
 
-def get_beta_mats_with_pow(F, pow=1):
+def normalize_matrix(m):
```

Comment:
What's this about?

Reply:
Matrices are not hashable (at least if they are mutable). I wanted to remove duplicates so I had to create a function that does this. In practice removing duplicates cuts the number of beta_mats to check in half. For the most time consuming step, namely type (d,1,...,1,1,...,1) this reduces the number of beta vectors by a factor 2^d so a significant speedup. Also if m is diagonal (happens sometimes for SS elliptic curves) then the bound computed with

```
x1 = [b 0]
     [0 b]
```
is just a square of the bound computed with:
```
x2 = [b]
```
The reason even after all the tensoring the pil_mat computed x1 (lets call it pil_mat_x1) will be a block diagonal matrix whose diagonal is pil_mat_x2 or in formulas:

```
pil_mat_x1 = [ pil_mat_x2          0 ]
             [          0 pil_mat_x2 ]
``` 
It is clear that det(pil_mat_x1) = det(pil_mat_x2)^2 so the prime divisors are the same.